src/librustdoc/clean/mod.rs

```diff
@@ -1479,8 +1479,88 @@ pub(crate) fn clean_middle_assoc_item<'tcx>(
     Item::from_def_id_and_parts(assoc_item.def_id, Some(assoc_item.name), kind, cx)
 }
 
+/// The goal of this function is to return the first `Path` which is not private (ie not private
+/// or `doc(hidden)`). If it's not possible, it'll return the "end type".
+///
+/// If the path is not a re-export or is public, it'll return `None`.
+fn first_non_private(
+    cx: &mut DocContext<'_>,
+    hir_id: hir::HirId,
+    path: &hir::Path<'_>,
+) -> Option<Path> {
+    let (parent_def_id, mut ident) = match &path.segments[..] {
+        [] => return None,
+        // Relative paths are available in the same scope as the owner.
+        [leaf] => (cx.tcx.local_parent(hir_id.owner.def_id), leaf.ident),
+        // So are self paths.
+        [parent, leaf] if parent.ident.name == kw::SelfLower => {
+            (cx.tcx.local_parent(hir_id.owner.def_id), leaf.ident)
+        }
+        // Crate paths are not. We start from the crate root.
+        [parent, leaf] if parent.ident.name == kw::Crate => {
+            (LOCAL_CRATE.as_def_id().as_local()?, leaf.ident)
+        }
```

Comment:
This also needs to handle `kw::Super`.

Reply:
Consider:

```rs
use external::Public as Private;

pub mod external {
   pub struct Public;

   pub fn make() -> super::Private { super::Private }
   //pub fn make() -> crate::Private { crate::Private }
}
```

Reply:
Good point. I'll add a test as well, good catch!

Reply:
As well as `kw::PathRoot` (`{{root}}`) for `::Private` in edition 2015.

```rs
// edition:2015
pub fn make() -> ::Private { ::Private }
```

Reply:
I think we can specify the edition we want in tests so should be simple.